### ISSUE TITLE
White-Box model results are not good

### ISSUE THREAD

**Issue by yxt132** [zhen8838/AnimeStylized#2]
Great work in re-implementing the white-box model using pytorch. However, after testing, I found the results are not good as the official version by the authors. There is still a large gap. What do you think could be the reason? Do we need the train the model longer or something else?

**Comment by zhen8838** [comment]
Yes, I also think the white-box model not as good as the official version. 

1. I found the re-implemented model predicts the result has a wild color(you can see right-top area). The original repo has the same [problem](https://github.com/SystemErrorWang/White-box-Cartoonization/issues/47).
![animegan_test2_out](https://user-images.githubusercontent.com/26156999/98645930-9b5e7b80-236d-11eb-8255-a2211127c644.jpg)

2. re-implemented model predicts anime-style is not obvious enough.

When I re-implementing, I spend a lot of time ensuring the `color shift` `gudied_filter` will have the same behavior. I think the training step is the same as the official version.

Now I try to use the same hyperparameters training both steps in the official version and my version. This model training costs a lot of time in superpixel, So I need some time to test.

If you can help find which code has a problem, I will be very grateful.

**Comment by yxt132** [comment]
thanks for your quick response! I have not started the training yet. I will let you know if I figure out something. By the way, which superpixel method did you use in your training. I wonder how much impact the superpixel method has on the results.

**Comment by zhen8838** [comment]
My default superpixel method during training is consistent with the one mentioned in the author’s paper. He uses the superpixel method of adaptive brightness to increase the brightness of the output image, and the parameters I use are consistent with the official code, `sigma=1.2 seg_num=200`, you can check my config file.

**Comment by zhen8838** [comment]
official code use `selective_adacolor` superpixel method training 15999 steps results：
![15999_face_photo](https://user-images.githubusercontent.com/26156999/98946281-a7de0200-252e-11eb-81e8-0e2c9dd3995e.jpg)
![15999_face_result](https://user-images.githubusercontent.com/26156999/98946288-aa405c00-252e-11eb-804b-958e039df0c8.jpg)
![15999_scenery_photo](https://user-images.githubusercontent.com/26156999/98946313-b2000080-252e-11eb-8877-3187c727752b.jpg)
![15999_scenery_result](https://user-images.githubusercontent.com/26156999/98946321-b3c9c400-252e-11eb-9a40-79307874c6e0.jpg)

**Comment by yxt132** [comment]
Not bad. I noticed some strange color in the generated pictures though.

![image](https://user-images.githubusercontent.com/43161328/99024674-07292a00-25a2-11eb-808c-227dc044975c.png)

how is the pytorch version's training? any progress?

**Comment by zhen8838** [comment]
|test images|official code|pytorch version|
|--|--|--|
|![actress2](https://user-images.githubusercontent.com/26156999/99029488-8c660c00-25ad-11eb-8440-3245a44a9d76.jpg)|![actress2](https://user-images.githubusercontent.com/26156999/99029726-2928a980-25ae-11eb-8575-69985d89b1e5.jpg)|![actress2_out](https://user-images.githubusercontent.com/26156999/99029955-d3083600-25ae-11eb-9290-d67b7b386f7a.jpg)|
|![china6](https://user-images.githubusercontent.com/26156999/99029533-abfd3480-25ad-11eb-98c4-f249b2addd7b.jpg)|![china6](https://user-images.githubusercontent.com/26156999/99029762-4f4e4980-25ae-11eb-9f45-6bee3c58f1c7.jpg)|![china6_out](https://user-images.githubusercontent.com/26156999/99029968-d996ad80-25ae-11eb-92b0-837080304aee.jpg)|
|![food6](https://user-images.githubusercontent.com/26156999/99029545-b3244280-25ad-11eb-98e4-dd352debbeef.jpg)|![food6](https://user-images.githubusercontent.com/26156999/99029769-55dcc100-25ae-11eb-8061-e24ad3a40f60.jpg)|![food6_out](https://user-images.githubusercontent.com/26156999/99029976-def3f800-25ae-11eb-843b-b078534da1b3.jpg)|
|![food16](https://user-images.githubusercontent.com/26156999/99029560-bdded780-25ad-11eb-9914-d2ad933c735d.jpg)|![food16](https://user-images.githubusercontent.com/26156999/99029790-5f662900-25ae-11eb-99dc-a1a216b17c5c.jpg)|![food16_out](https://user-images.githubusercontent.com/26156999/99029992-e7e4c980-25ae-11eb-86fb-8c2a1ff0fca5.jpg)|
|![liuyifei4](https://user-images.githubusercontent.com/26156999/99029574-c505e580-25ad-11eb-8cf2-6d539775821d.jpg)|![liuyifei4](https://user-images.githubusercontent.com/26156999/99029807-69882780-25ae-11eb-8db2-6c5b8019053c.jpg)|![liuyifei4_out](https://user-images.githubusercontent.com/26156999/99029999-ed421400-25ae-11eb-97a9-c547ba80221a.jpg)|
|![london1](https://user-images.githubusercontent.com/26156999/99029662-00081900-25ae-11eb-9a13-148293575cfc.jpg)|![london1](https://user-images.githubusercontent.com/26156999/99029816-70169f00-25ae-11eb-832e-e5cd40ffd260.jpg)|![london1_out](https://user-images.githubusercontent.com/26156999/99030014-f6cb7c00-25ae-11eb-9de6-61afd6a7c146.jpg)|
|![mountain4](https://user-images.githubusercontent.com/26156999/99029665-01d1dc80-25ae-11eb-9d1a-50e642e1ffa2.jpg)|![mountain4](https://user-images.githubusercontent.com/26156999/99029821-760c8000-25ae-11eb-9306-35a71f729787.jpg)|![mountain4_out](https://user-images.githubusercontent.com/26156999/99030023-faf79980-25ae-11eb-86f8-1686b1066de0.jpg)|
|![mountain5](https://user-images.githubusercontent.com/26156999/99029669-04343680-25ae-11eb-81ec-f43ef41d4b73.jpg)|![mountain5](https://user-images.githubusercontent.com/26156999/99029846-815fab80-25ae-11eb-9928-d7698d72d47b.jpg)|![mountain5_out](https://user-images.githubusercontent.com/26156999/99030034-02b73e00-25af-11eb-8e2c-6701207872cb.jpg)|
|![national_park1](https://user-images.githubusercontent.com/26156999/99029672-05fdfa00-25ae-11eb-9826-c893204518e1.jpg)|![national_park1](https://user-images.githubusercontent.com/26156999/99029853-86245f80-25ae-11eb-9f22-ebbf667124b8.jpg)|![national_park1_out](https://user-images.githubusercontent.com/26156999/99030050-0a76e280-25af-11eb-962e-fcfa8f498fde.jpg)|
|![party5](https://user-images.githubusercontent.com/26156999/99029675-072f2700-25ae-11eb-99f7-12a6e3eb37a0.jpg)|![party5](https://user-images.githubusercontent.com/26156999/99029864-94727b80-25ae-11eb-866d-1cc251a93ff7.jpg)|![party5_out](https://user-images.githubusercontent.com/26156999/99030058-0f3b9680-25af-11eb-9904-e3ad41986498.jpg)|
|![party7](https://user-images.githubusercontent.com/26156999/99029680-08605400-25ae-11eb-8ed2-988a149ace78.jpg)|![party7](https://user-images.githubusercontent.com/26156999/99029881-9b998980-25ae-11eb-9287-298a595d36cb.jpg)|![party7_out](https://user-images.githubusercontent.com/26156999/99030070-15317780-25af-11eb-89cf-f04a92bb6960.jpg)|

**Comment by yxt132** [comment]
Well done!  It seems the pytorch version's results are smoother than the official tf version. What changes did you make in the pytorch version? I acutally like the pytorch version's results better. Can you update your repo and release the updated trained weights? Again, great work!

**Comment by zhen8838** [comment]
Hi. I add new weights in google drive, you can find in the readme.  I also upload tensorflow version weights named `whitebox-tf.zip`.

**Comment by zhen8838** [comment]
I found the strange color caused by `guided filter`, but now I didn't find a better method to solve it.

**Comment by yxt132** [comment]
the author said you could train without guided filter and add guided filter during inference.

**Comment by zhen8838** [comment]
ok. I will try if time permits

**Comment by GustavoStahl** [comment]
One thing that you could try with the colors is to use **Color Transfer** algorithm, like [this](https://www.pyimagesearch.com/2014/06/30/super-fast-color-transfer-images/) from PyImageSearch.

Also, regarding cartoon noise, the guided filter should help on the post-process, by using lower values of epsilon (ε), like in WhiteBox's [cartoonize.py](https://github.com/SystemErrorWang/White-box-Cartoonization/blob/50b115e5adfefa12e0aad7b128a9e7cce7a9f2c5/test_code/cartoonize.py#L28).

**Comment by zhen8838** [comment]
@GustavoStahl  thanks, I missing the test_code ε value is not equal to train_code ε. But I don't think the *color transfer algorithm* is needed. For this model, it needs to keep the original color as much as possible and only increase the brightness. Regarding the degree of texture, I think it can be adjusted g_gray_weight.  like [this](https://github.com/SystemErrorWang/White-box-Cartoonization/issues/18).

**Comment by huangfuyang** [comment]
> I found the strange color caused by `guided filter`, but now I didn't find a better method to solve it.

adding np.clipping() after guided filter would solve the artifacts
